# 1. Heart Failure Prediction Dataset

In [58]:
import numpy as np
import pandas as pd

In [59]:
#1.1a
heart_df = pd.read_csv('datasets/heart_failure_dataset.csv')
print(heart_df.shape)


(299, 13)


In [60]:
print(heart_df.axes)


[RangeIndex(start=0, stop=299, step=1), Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')]


In [61]:
print(heart_df.dtypes)


age                         float64
anaemia                       int64
creatinine_phosphokinase      int64
diabetes                      int64
ejection_fraction             int64
high_blood_pressure           int64
platelets                   float64
serum_creatinine            float64
serum_sodium                  int64
sex                           int64
smoking                       int64
time                          int64
DEATH_EVENT                   int64
dtype: object


In [62]:
#1.1b
heart_df.head(10)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   
5  90.0        1                        47         0                 40   
6  75.0        1                       246         0                 15   
7  60.0        1                       315         1                 60   
8  65.0        0                       157         0                 65   
9  80.0        1                       123         0                 35   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   
5                    1  204000.00               2.1           132    1   
6                    0  127000.00               1.2           137    1   
7                    0  454000.00               1.1           131    1   
8                    0  263358.03               1.5           138    0   
9                    1  388000.00               9.4           133    1   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1  
5        1     8            1  
6        0    10            1  
7        1    10            1  
8        0    10            1  
9        1    10            1

**1.1c** Each row of heart_df dataframe can be termed an **instance** within the machine learning parlance.

**1.1d** We are dealing with a supervised learning problem. This is because our dataset has a distinct labelled output, DEATH_EVENT. We are attempting to find relationships between the input and output variables to evaluate whether a person will die due to heart failure.


**1.1e** This dataset has 12 independent variables. These variables are:
1. age
2. anaemia
3. creatinine_phosphokinase
4. diabetes
5. ejection_fraction
6. high_blood_pressure
7. platelets
8. serum_creatinine
9. serum_sodium
10. sex
11. smoking
12. time


**1.1f** This dataset has only 1 dependent variable.
- DEATH_EVENT

**1.1g** This dataset contains 6 categorical variables. These variables are: 
1. anaemia
2. diabetes
3. high_blood_pressure
4. sex
5. smoking
6. DEATH_EVENT

**1.1h** The cateforical variables in the dataset Ordinal encoding.

In [63]:
#1.1j
print("Deceased:", len(heart_df[heart_df.DEATH_EVENT == 1]))
print("Remaining", len(heart_df[heart_df.DEATH_EVENT == 0]))


Deceased: 96
Remaining 203


There are 96 deceased patients and 203 patients remaining.

In [64]:
#1.1j
print("Women:", len(heart_df[heart_df.sex == 0]))
print("Male:", len(heart_df[heart_df.sex == 1]))


Women: 105
Male: 194


There are 105 women patients and 194 male patients.

In [65]:
#1.1k
print("Smokers:", len(heart_df[heart_df.smoking == 1]))
print("Non-smokers:", len(heart_df[heart_df.smoking == 0]))


Smokers: 96
Non-smokers: 203


There are 96 smokers and 203 non-smokers.

In [66]:
#1.2
heart_df.corr()

age   anaemia  creatinine_phosphokinase  \
age                       1.000000  0.088006                 -0.081584   
anaemia                   0.088006  1.000000                 -0.190741   
creatinine_phosphokinase -0.081584 -0.190741                  1.000000   
diabetes                 -0.101012 -0.012729                 -0.009639   
ejection_fraction         0.060098  0.031557                 -0.044080   
high_blood_pressure       0.093289  0.038182                 -0.070590   
platelets                -0.052354 -0.043786                  0.024463   
serum_creatinine          0.159187  0.052174                 -0.016408   
serum_sodium             -0.045966  0.041882                  0.059550   
sex                       0.065430 -0.094769                  0.079791   
smoking                   0.018668 -0.107290                  0.002421   
time                     -0.224068 -0.141414                 -0.009346   
DEATH_EVENT               0.253729  0.066270                  0.062728   

                          diabetes  ejection_fraction  high_blood_pressure  \
age                      -0.101012           0.060098             0.093289   
anaemia                  -0.012729           0.031557             0.038182   
creatinine_phosphokinase -0.009639          -0.044080            -0.070590   
diabetes                  1.000000          -0.004850            -0.012732   
ejection_fraction        -0.004850           1.000000             0.024445   
high_blood_pressure      -0.012732           0.024445             1.000000   
platelets                 0.092193           0.072177             0.049963   
serum_creatinine         -0.046975          -0.011302            -0.004935   
serum_sodium             -0.089551           0.175902             0.037109   
sex                      -0.157730          -0.148386            -0.104615   
smoking                  -0.147173          -0.067315            -0.055711   
time                      0.033726           0.041729            -0.196439   
DEATH_EVENT              -0.001943          -0.268603             0.079351   

                          platelets  serum_creatinine  serum_sodium       sex  \
age                       -0.052354          0.159187     -0.045966  0.065430   
anaemia                   -0.043786          0.052174      0.041882 -0.094769   
creatinine_phosphokinase   0.024463         -0.016408      0.059550  0.079791   
diabetes                   0.092193         -0.046975     -0.089551 -0.157730   
ejection_fraction          0.072177         -0.011302      0.175902 -0.148386   
high_blood_pressure        0.049963         -0.004935      0.037109 -0.104615   
platelets                  1.000000         -0.041198      0.062125 -0.125120   
serum_creatinine          -0.041198          1.000000     -0.189095  0.006970   
serum_sodium               0.062125         -0.189095      1.000000 -0.027566   
sex                       -0.125120          0.006970     -0.027566  1.000000   
smoking                    0.028234         -0.027414      0.004813  0.445892   
time                       0.010514         -0.149315      0.087640 -0.015608   
DEATH_EVENT               -0.049139          0.294278     -0.195204 -0.004316   

                           smoking      time  DEATH_EVENT  
age                       0.018668 -0.224068     0.253729  
anaemia                  -0.107290 -0.141414     0.066270  
creatinine_phosphokinase  0.002421 -0.009346     0.062728  
diabetes                 -0.147173  0.033726    -0.001943  
ejection_fraction        -0.067315  0.041729    -0.268603  
high_blood_pressure      -0.055711 -0.196439     0.079351  
platelets                 0.028234  0.010514    -0.049139  
serum_creatinine         -0.027414 -0.149315     0.294278  
serum_sodium              0.004813  0.087640    -0.195204  
sex                       0.445892 -0.015608    -0.004316  
smoking                   1.000000 -0.022839    -0.012623  
time                     -0.022839  1.000000  

**1.2a** Age and serum_creatinine are the two variables most positively correlated with the DEATH_EVENT variable.

**1.2b** Ejection_fraction and time are the two variables most negatively correlated with the DEATH_EVENT variable.

**1.2c** The second-most positively correlated variable with the DEATH_EVENT variable is Age. It makes sense that these two variables are positively correlated because as you get older, the rate of death by heart failure increases.

**1.2d** The two variables most negatively correlated with DEATH_EVENT are Time and Ejection_fraction. <br>
Time being negatively correlated with DEATH_EVENT makes sense because the follow-time is likely shorter if a given patient passes. <br>
Ejection_fraction being negatively correlated with DEATH_EVENT also makes sense because a the lower the fraction of blood a heart ejects, the unhealthier or weaker it may be, leading to an increased chance of heart failure and therefore the DEATH_EVENT. On the other hand, a healthier heart will likely eject larger fractions of blood, decreasing the chance of death through heart failure.

**1.3** To validate this dataset, we can check cariables for null entries. We can check for zeroes and whiespace entries on the following numerical variables:
1. age
2. creatinine_phosphokinase
3. ejection_fraction
4. platelets
5. serum_creatinine
6. serum_sodium

<br> 
The other variables are either categorical in nature and cannot benefit from zero-checking, or are numerical with the possibility of zero, such as time with a possible same-day follow up. 
<br>
We can also check all data entries for non-numeric and negative entries.
<br>
To finish off, we can check all data entries for invalid null, or whitespace, entries.

In [67]:
#Replace all zeroes in age, creatinine_phosphokinase, ejection_fraction, platelets, serum_creatinine, and serum_sodium with NaN.
heart_df['age'].replace(0, np.nan, inplace = True)
heart_df['creatinine_phosphokinase'].replace(0, np.nan, inplace = True)
heart_df['ejection_fraction'].replace(0, np.nan, inplace = True)
heart_df['platelets'].replace(0, np.nan, inplace = True)
heart_df['serum_creatinine'].replace(0, np.nan, inplace = True)
heart_df['serum_sodium'].replace(0, np.nan, inplace = True)

#Replace all non-numeric entries with NaN.
heart_df.replace(r'[a-zA-Z]', np.nan, inplace = True, regex = True)

#Replace all negative entries with NaN.
heart_df.replace(r'-', np.nan, inplace = True, regex = True)

#Replace all whitespaces in entire dataset with NaN.
heart_df.replace(r'^\s*$', np.nan, inplace = True, regex=True)

#Export to csv
heart_df.to_csv('datasets/heart_failure_dataset_NaNs.csv', index=False)

In [68]:
#1.4
#Defining the non-categorical variables
heart_numerical = ["age", "creatinine_phosphokinase", "ejection_fraction", "platelets", "serum_creatinine", 
                   "serum_sodium", "time"]

#Normalizing the data using (df-df.mean())/df.std()
#We set skipna to be true, so we can ignore NaN entries
#We include an if statement at the end to ensure we are only processing the non-categorical independent variables.
heart_df = heart_df.apply(lambda x: (x - x.mean(skipna = True))/x.std(skipna = True) if x.name in heart_numerical else x)

In [69]:
print(heart_df.head(20))

         age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0   1.190949        0                  0.000165         0          -1.527998   
1  -0.490457        0                  7.502063         0          -0.007065   
2   0.350246        0                 -0.449186         0          -1.527998   
3  -0.910808        1                 -0.485257         0          -1.527998   
4   0.350246        1                 -0.434757         1          -1.527998   
5   2.452003        1                 -0.551217         0           0.161928   
6   1.190949        1                 -0.346124         0          -1.950479   
7  -0.070106        1                 -0.275011         1           1.851853   
8   0.350246        0                 -0.437849         0           2.274335   
9   1.611300        1                 -0.472890         0          -0.260554   
10  1.190949        1                 -0.516176         0          -0.007065   
11  0.098035        0                 -0

**1.5** To utilize one-hot encoding, we can process the DEATH_EVENT variable and output two dependent variables: Remaining and Deceased. To do this, we can create a new column for Remaining, and we can reuse the DEATH_EVENT column for our Deceased column.

In [70]:
#Create new Remaining column based off of when DEATH_EVENT is zero
heart_df['Remaining'] = (heart_df['DEATH_EVENT'] == 0) * 1

#Reuse the DEATH_EVENT column as our Deceased column
heart_df.rename(columns = {'DEATH_EVENT' : 'Deceased'}, inplace = True)

#Export to csv
heart_df.to_csv('datasets/heart_failure_dataset_processed.csv', index = False)

# 2. Pima Indians Diabetes Dataset

In [71]:
diabetes_df = pd.read_csv('datasets/diabetes_dataset.csv')


In [72]:
#2.1a
print(diabetes_df.shape)
print(diabetes_df.axes)
print(diabetes_df.dtypes)

(768, 9)
[RangeIndex(start=0, stop=768, step=1), Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')]
Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


In [73]:
#2.1b
print(diabetes_df.head(10))

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   2

**2.1c** We are dealing with a supervised learning problem. This is because our dataset has a distinct labelled output, outcome. We are attempting to find relationships between the input and output variables to evaluate whether or not a person has diabetes.

**2.1d** This dataset has 8 independent variables. These variables are:
1. Pregnancies
2. Glucose
3. BloodPressure
4. SkinThickness
5. Insulin
6. BMI
7. DiabetesPedigreeFunction
8. Age


**2.1e** This dataset has only 1 dependent variable.
- Outcome

**2.1f** This dataset has only 1 categorical variable.
- Outcome

**2.1g** This categorical variable uses ordinal encoding.

In [74]:
#2.1h
print("Young adults:", len(diabetes_df[(diabetes_df.Age >= 21) & (diabetes_df.Age <= 40)]))
print("Middle-aged adults:", len(diabetes_df[(diabetes_df.Age >= 41) & (diabetes_df.Age <= 60)]))
print("Old-aged adults:", len(diabetes_df[(diabetes_df.Age >= 61)]))


Young adults: 574
Middle-aged adults: 167
Old-aged adults: 27


There are 574 Young adults, 167 Middle-aged adults, and 27 Old-aged adults.

**2.2** To validate this dataset, we can check cariables for null entries. We can check for zero entries on the following numerical variables:
1. Glucose
2. BloodPressure
3. SkinThickness
4. Insulin
5. BMI
6. DiabetesPedigreeFunction
7. Age

<br> 
The other variable is Pregnancies, where a value of zero is valid.
<br>
We can also check all data entries for non-numeric and negative entries.
<br>
To finish off, we can check all data entries for invalid null, or whitespace, entries.

In [75]:
#Replace all zeroes in Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, and Age with NaN.
diabetes_df['Glucose'].replace(0, np.nan, inplace = True)
diabetes_df['BloodPressure'].replace(0, np.nan, inplace = True)
diabetes_df['SkinThickness'].replace(0, np.nan, inplace = True)
diabetes_df['Insulin'].replace(0, np.nan, inplace = True)
diabetes_df['BMI'].replace(0, np.nan, inplace = True)
diabetes_df['DiabetesPedigreeFunction'].replace(0, np.nan, inplace = True)
diabetes_df['Age'].replace(0, np.nan, inplace = True)

#Replace all non-numeric entries with NaN.
diabetes_df.replace(r'[a-zA-Z]', np.nan, inplace = True, regex = True)

#Replace all negative entries with NaN.
diabetes_df.replace(r'-', np.nan, inplace = True, regex = True)

#Replace all whitespaces in entire dataset with NaN.
diabetes_df.replace(r'^\s*$', np.nan, inplace = True, regex=True)

#Print
print(diabetes_df.head(20))

    Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             6    148.0           72.0           35.0      NaN  33.6   
1             1     85.0           66.0           29.0      NaN  26.6   
2             8    183.0           64.0            NaN      NaN  23.3   
3             1     89.0           66.0           23.0     94.0  28.1   
4             0    137.0           40.0           35.0    168.0  43.1   
5             5    116.0           74.0            NaN      NaN  25.6   
6             3     78.0           50.0           32.0     88.0  31.0   
7            10    115.0            NaN            NaN      NaN  35.3   
8             2    197.0           70.0           45.0    543.0  30.5   
9             8    125.0           96.0            NaN      NaN   NaN   
10            4    110.0           92.0            NaN      NaN  37.6   
11           10    168.0           74.0            NaN      NaN  38.0   
12           10    139.0           80.0            

In [76]:
#2.3
#Defining the non-categorical variables
diabetes_numerical = ["Pregnancies", "Glucose", "BloodPressure", "SkinThickness", "Insulin", 
                      "BMI", "DiabetesPedigreeFunction", "Age"]

#Normalizing the data using (df-df.mean())/df.std()
#We set skipna to be true, so we can ignore NaN entries
#We include an if statement at the end to ensure we are only processing the non-categorical independent variables.
diabetes_df = diabetes_df.apply(lambda x: (x - x.mean(skipna = True))/x.std(skipna = True) if x.name in diabetes_numerical else x)

#Print
print(diabetes_df.head(20))

    Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0      0.639530  0.861722      -0.032723       0.558040       NaN  0.164987   
1     -0.844335 -1.201441      -0.517291      -0.014643       NaN -0.845845   
2      1.233077  2.007924      -0.678814            NaN       NaN -1.322380   
3     -0.844335 -1.070446      -0.517291      -0.587327 -0.518188 -0.629238   
4     -1.141108  0.501487      -2.617087       0.558040  0.104834  1.536831   
5      0.342757 -0.186234       0.128800            NaN       NaN -0.990249   
6     -0.250789 -1.430681      -1.809473       0.271699 -0.568703 -0.210464   
7      1.826623 -0.218982            NaN            NaN       NaN  0.410475   
8     -0.547562  2.466404      -0.194246       1.512514  3.262042 -0.282667   
9      1.233077  0.108504       1.905550            NaN       NaN       NaN   
10     0.045984 -0.382725       1.582504            NaN       NaN  0.742606   
11     1.826623  1.516694       0.128800            

**2.4** To utilize one-hot encoding, we can process the Outcome variable and output two dependent variables: HasDiabetes and NoDiabetes. To do this, we can create a new column for NoDiabetes, and we can reuse the Outcome column for our Deceased column.

In [77]:
#Create new Remaining column based off of when DEATH_EVENT is zero
diabetes_df['NoDiabetes'] = (diabetes_df['Outcome'] == 0) * 1

#Reuse the DEATH_EVENT column as our Deceased column
diabetes_df.rename(columns = {'Outcome' : 'HasDiabetes'}, inplace = True)


In [78]:
#2.5
#Create new dataframe that is grouped by the dependent variable HasDiabetes.
#This variable is equivalent to Outcome.
#This new dataframe, median_df, contains median of each variable depending on Outcome (0, 1)
median_df = diabetes_df.groupby('HasDiabetes').transform(np.median)
#Replaces all null spaces (NaN) with the median for that value, value stored in median_df
diabetes_df[diabetes_df.isnull()] = median_df

#Export to csv
diabetes_df.to_csv('datasets/diabetes_dataset_processed.csv', index = False)